src/server/mail/read.ts

```diff
@@ -0,0 +1,484 @@
+import { userFilterSelection } from '@/server/users/ConfigVars'
+import { ServerError } from '@/server/error'
+import prisma from '@/prisma'
+import type { MailFlowObject, MailListTypes, ViaType } from './Types'
+
+
+/**
+ * This file reads how incomming mail is routed.
+ * Mail flows in one direction
+ * Mailalis -> Mailinglist  -> User
+ *                          -> Group -> User
+ *                          -> External address
+ * 
+ * To show how the mail traverses, each node in the network above can be used as a source to read the traversal in both directions.
+ * Each of the traversal functions is divieded in four parts
+ *  1. Fetch the data from the database
+ *  2. Parse the data into arrays with only one type of object
+ *       This step also adds a via object (next paragraph) if it is nessasary
+ *  3. Delete unnessesary fields
+ *  4. Combine duplicate nodes
+ * 
+ * 
+ * Since some nodes may have different paths to itself, a via property is used.
+ * For example a user can be directly connected to a mailinglist or via a group.
+ * We only want the user to appear once, but we want to return all the paths to the users (both via the group and direct).
+ * 
+ * 
+ * The end results should be an object that can be displayed to the user to show how mail is routed though a arbirtrary node in the network.
+*/
+
+/**
+ * Removes duplicates from an array of objects based on their `id` property.
+ * If multiple objects have the same `id`, the first object encountered is kept,
+ * and the `via` property of subsequent objects with the same `id` is merged into the first object's `via` array.
+ * @param objects An array of objects to remove duplicates from.
+ * @returns An array of objects with duplicates removed.
+ */
+function removeDuplicates<T extends {
+    id: number,
+} & ViaType>(objects: T[]): T[] {
+    const seenIds = new Set()
+    const ret: T[] = []
+
+    for (const obj of objects) {
+        if (seenIds.has(obj.id)) {
+            const exisingObj = ret.find(item => item.id === obj.id)
+            if (exisingObj && exisingObj.via && obj.via && obj.via instanceof Array) {
+                const via = obj.via[0]
+
+                const alredyAdded = exisingObj.via.find(item => (item.type === via.type && item.id === via.id))
+                if (!alredyAdded) {
+                    exisingObj.via.push(via)
+                }
+            }
+        } else {
+            seenIds.add(obj.id)
+            ret.push(obj)
+        }
+    }
+
+    return ret
+}
+
+/**
+ * Reads the mail flow based on a mailalias with the provided ID.
+ * @param id - The ID of the mail alias.
+ * @returns A Promise that resolves to a MailFlowObject.
+ */
+async function readAliasTraversal(id: number): Promise<MailFlowObject> {
+    const results = await prismaCall(() => prisma.mailAlias.findUniqueOrThrow({
+        where: {
+            id,
+        },
+        include: {
+            mailingLists: {
+                include: {
```

Comment:
This should have some includers. Often (look at cms modules) when we had long includes I started aking them in ConfigVars.

Reply:
I can move the include statements to a ConfigVars file, the the include statements will probably only be used one place.